subsys/bluetooth/host/conn.c

```diff
@@ -614,429 +592,460 @@ static inline uint16_t conn_mtu(struct bt_conn *conn)
- *   and `buf` is still the head of the TX queue
+ * Partitions the controller's buffers to each connection according to some
+ * heuristic. This is made to be tunable, fairness, simplicity, throughput etc.
  *
+ * In the future, this will be a hook exposed to the application.
  */
-static int send_buf(struct bt_conn *conn, struct net_buf *buf)
+bool bt_data_should_stop(struct bt_conn *conn)
 {
-	struct net_buf *frag;
-	uint8_t flags;
-	int err;
+	LOG_DBG("%p", conn);
 
-	LOG_DBG("conn %p buf %p len %u", conn, buf, buf->len);
-
-	/* Send directly if the packet fits the ACL MTU */
-	if (buf->len <= conn_mtu(conn) && !tx_data(buf)->is_cont) {
-		LOG_DBG("send single");
-		return send_frag(conn, buf, NULL, FRAG_SINGLE);
-	}
-
-	LOG_DBG("start fragmenting");
-	/*
-	 * Send the fragments. For the last one simply use the original
-	 * buffer (which works since we've used net_buf_pull on it).
+	/* TODO: This function should be overridable by the application: they
+	 * should be able to provide their own heuristic.
 	 */
-	flags = FRAG_START;
-	if (tx_data(buf)->is_cont) {
-		flags = FRAG_CONT;
+	if (!conn->has_data(conn)) {
+		LOG_DBG("No more data for %p", conn);
+		return true;
 	}
 
-	while (buf->len > conn_mtu(conn)) {
-		frag = create_frag(conn, buf);
-		if (!frag) {
-			return -ENOMEM;
-		}
-
-		err = send_frag(conn, buf, frag, flags);
-		if (err) {
-			LOG_DBG("%p failed, mark as existing frag", buf);
-			tx_data(buf)->is_cont = flags != FRAG_START;
-			net_buf_unref(frag);
-			return err;
-		}
-
-		flags = FRAG_CONT;
+	/* Queue only 3 buffers per-conn for now */
+	if (atomic_get(&conn->in_ll) < 3) {
+		/* The goal of this heuristic is to allow the link-layer to
+		 * extend an ACL connection event as long as the application
+		 * layer can provide data.
+		 *
+		 * Here we chose three buffers, as some LLs need two enqueued
+		 * packets to be able to set the more-data bit, and one more
+		 * buffer to allow refilling by the app while one of them is
+		 * being sent over-the-air.
+		 */
+		return false;
 	}
 
-	LOG_DBG("last frag");
-	tx_data(buf)->is_cont = true;
-	return send_frag(conn, buf, NULL, FRAG_END);
+	return true;
 }
 
-static struct k_poll_signal conn_change =
-		K_POLL_SIGNAL_INITIALIZER(conn_change);
-
-static void conn_cleanup(struct bt_conn *conn)
+void bt_conn_data_ready(struct bt_conn *conn)
 {
-	struct net_buf *buf;
-
-	/* Give back any allocated buffers */
-	while ((buf = net_buf_get(&conn->tx_queue, K_NO_WAIT))) {
-		struct bt_conn_tx *tx = tx_data(buf)->tx;
+	LOG_DBG("DR");
 
-		tx_data(buf)->tx = NULL;
-
-		/* destroy the buffer */
-		net_buf_unref(buf);
-
-		/* destroy the tx context (and any associated meta-data) */
-		if (tx) {
-			conn_tx_destroy(conn, tx);
-		}
+	/* The TX processor will call the `pull_cb` to get the buf */
+	if (!atomic_set(&conn->_conn_ready_lock, 1)) {
+		sys_slist_append(&bt_dev.le.conn_ready,
+				 &conn->_conn_ready);
+		LOG_DBG("raised");
+	} else {
+		LOG_DBG("already in list");
```

Comment:
Suggested this before: IMO it makes sense to start all log statements with capital letters, as they are (for most part) sentences.

Reply:
What about `.` at the end? I've been yelled at before for including them

Reply:
That I don't have an opinion on to be honest. I typically add it if the log statement is more than one sentence, otherwise exclude it. (arguably if you log statement is more than 1 sentence, the author may need to think twice about it).